File: engine/ai-pilot.js
// Boucle de pilotage d'une étape « tâche IA » (Lot 5).
//
// engine/ conduit le navigateur ; Django décide. À chaque tour : résumer la page
// → demander l'action à Django (qui appelle le fournisseur IA avec la clé de
// l'utilisateur) → exécuter → recommencer, jusqu'à `finish`, `fail`, ou le
// plafond d'itérations.

import { selectorFor, summarize } from './page-summary.js';

const BACKEND_URL = process.env.BACKEND_INTERNAL_URL || 'http://robot-lab-backend:8000';
const ENGINE_KEY = process.env.ENGINE_INTERNAL_KEY || '';
const ACTION_TIMEOUT_MS = 15000;
// Un appel LLM est lent (secondes), mais pas illimité : sans plafond une session
// bloquée retiendrait le mutex de engine/ indéfiniment.
const DECISION_TIMEOUT_MS = 120000;

async function askBackend(payload) {
  const res = await fetch(`${BACKEND_URL}/api/internal/ai-step/`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Engine-Key': ENGINE_KEY },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(DECISION_TIMEOUT_MS),
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || `Décision IA refusée (${res.status})`);
  }
  return res.json();
}

/** Exécute l'action choisie. Renvoie un compte rendu court pour l'historique. */
async function applyAction(page, action) {
  const { tool, args } = action;
  const options = { timeout: ACTION_TIMEOUT_MS };

  switch (tool) {
    case 'click':
      await page.click(selectorFor(args.index), options);
      return 'clic effectué';
    case 'fill':
      await page.fill(selectorFor(args.index), args.value, options);
      return 'champ rempli';
    case 'select':
      await page.selectOption(selectorFor(args.index), args.value, options);
      return 'option choisie';
    case 'press':
      await page.press(selectorFor(args.index), args.key, options);
      return `touche ${args.key} envoyée`;
    case 'navigate':
      await page.goto(args.url, { waitUntil: 'domcontentloaded', ...options });
      return 'navigation effectuée';
    default:
      throw new Error(`Outil non exécutable : ${tool}`);
  }
}

/** Décrit une action en français — repris tel quel dans le journal du frontend. */
function describeAction(action, page) {
  const { tool, args } = action;
  const label = (i) => {
    const el = (page.elements || []).find(e => e.index === i);
    return el?.label ? `« ${el.label} »` : `élément ${i}`;
  };
  switch (tool) {
    case 'click':    return `Cliquer ${label(args.index)}`;
    case 'fill':     return `Remplir ${label(args.index)} avec « ${args.value} »`;
    case 'select':    return `Choisir « ${args.value} » dans ${label(args.index)}`;
    case 'press':    return `Appuyer sur ${args.key} dans ${label(args.index)}`;
    case 'navigate': return `Aller à ${args.url}`;
    case 'finish':   return `Objectif atteint : ${args.result}`;
    case 'fail':     return `Abandon : ${args.reason}`;
    default:         return tool;
  }
}

/**
 * Pilote une étape `ai_task`. `report(payload)` remonte chaque tour au client.
 * Résout `{ok, note}` — ne lève que sur erreur d'infrastructure.
 */
export async function runAiTask(page, step, { runId, maxIterations, report }) {
  const history = [];
  const cap = maxIterations || 12;

  for (let turn = 1; turn <= cap; turn++) {
    const summary = await summarize(page);

    let action;
    try {
      action = await askBackend({
        run_id: runId,
        objective: step.objective,
        expected_result: step.expected_result || '',
        page: summary,
        history,
      });
    } catch (err) {
      return { ok: false, note: `Décision IA impossible : ${err.message || err}` };
    }

    const label = describeAction(action, summary);
    report({ type: 'ai_action', turn, total: cap, label, reasoning: action.reasoning });

    if (action.tool === 'finish') {
      return { ok: true, note: action.args.result };
    }
    if (action.tool === 'fail') {
      return { ok: false, note: action.args.reason };
    }

    let outcome;
    try {
      outcome = await applyAction(page, action);
    } catch (err) {
      // L'échec n'arrête pas la tâche : il entre dans l'historique pour que l'IA
      // tente autrement au tour suivant. C'est le cœur de l'intérêt du mode IA
      // sur une page qui change — s'arrêter au premier raté le viderait de sens.
      outcome = `échec (${String(err.message || err).split('\n')[0]})`;
    }
    history.push({ summary: label, outcome });

    // Laisse la page réagir (navigation, requête XHR, rendu).
    await page.waitForTimeout(800);
  }

  return {
    ok: false,
    note: `Objectif non atteint après ${cap} tours (plafond de sécurité).`,
  };
}

File: backend/api/ai_pilot.py
"""Pilotage autonome d'une étape « tâche IA » (Lot 5).

Découpage : `engine/` conduit le navigateur et décrit la page ; **Django décide**
de l'action suivante. La clé API de l'utilisateur reste donc là où elle est déjà
déchiffrée (ici), et n'entre jamais dans le conteneur `engine/`.

L'IA ne fabrique jamais de sélecteur : `engine/` numérote les éléments
interactifs de la page et l'IA choisit un **index**. C'est ce qui évite le mode
d'échec bien connu du sélecteur halluciné (cf. `_valide()` de `restauration`,
qui doit revalider chaque XPath proposé par le LLM).

Le choix d'action passe par une sortie structurée JSON plutôt que par le
tool-calling natif : Claude et Mistral n'ont pas la même API d'outils, alors que
`ai_client.complete_json()` marche à l'identique sur les deux.
"""
from . import ai_client

MAX_ITERATIONS = 12

# outil → champs attendus dans `args`
TOOLS = {
    'click':    {'index'},
    'fill':     {'index', 'value'},
    'select':   {'index', 'value'},
    'press':    {'index', 'key'},
    'navigate': {'url'},
    'finish':   {'result'},
    'fail':     {'reason'},
}

SYSTEM = """\
Tu pilotes un navigateur pour accomplir un objectif précis sur une page web.

À chaque tour tu reçois : l'objectif, l'état de la page (titre, URL, texte
visible) et la liste numérotée de ses éléments interactifs. Tu choisis UNE action.

Tu ne peux agir que sur les éléments listés, en donnant leur numéro (`index`).
N'invente jamais de numéro absent de la liste, ni de sélecteur CSS.

Outils disponibles :
- `click` {index} — cliquer l'élément.
- `fill` {index, value} — saisir une valeur dans un champ.
- `select` {index, value} — choisir une option dans une liste déroulante.
- `press` {index, key} — appuyer une touche (`Enter`, `Tab`…) dans un élément.
- `navigate` {url} — aller directement à une URL.
- `finish` {result} — l'objectif est atteint. `result` décrit ce qui a été
  obtenu (et les informations extraites, s'il y en a).
- `fail` {reason} — l'objectif est impossible ici. Explique pourquoi.

Règles :
- Une seule action par tour. Après chaque action tu reverras la page mise à jour.
- Regarde l'historique : ne répète pas une action qui vient d'échouer, essaie
  autrement.
- Ne remplis jamais un champ de mot de passe : ces valeurs ne sont pas fournies.
- Appelle `finish` dès que le résultat attendu est visible — ne continue pas
  « pour vérifier ».
- Appelle `fail` plutôt que de tourner en rond si la page ne permet pas
  d'atteindre l'objectif.

Réponds avec `tool`, `args`, et `reasoning` (une phrase courte en français
expliquant ton choix — elle est affichée à l'utilisateur).
"""

ACTION_SCHEMA = {
    'type': 'object',
    'properties': {
        'tool': {'type': 'string', 'enum': sorted(TOOLS)},
        'args': {
            'type': 'object',
            'properties': {
                'index': {'type': 'integer'},
                'value': {'type': 'string'},
                'key': {'type': 'string'},
                'url': {'type': 'string'},
                'result': {'type': 'string'},
                'reason': {'type': 'string'},
            },
            'additionalProperties': False,
        },
        'reasoning': {'type': 'string'},
    },
    'required': ['tool', 'args', 'reasoning'],
    'additionalProperties': False,
}


class PilotError(Exception):
    """Action inutilisable — message destiné à l'utilisateur."""


def _render_page(page):
    """Met la page en texte pour le LLM. `page` vient de `engine/page-summary.js`."""
    lines = [
        f"Titre : {page.get('title') or '(sans titre)'}",
        f"URL : {page.get('url') or '(inconnue)'}",
    ]

    elements = page.get('elements') or []
    if elements:
        lines.append('\nÉléments interactifs :')
        for el in elements:
            label = el.get('label') or '(sans libellé)'
            detail = f" role={el.get('role')}"
            if el.get('type'):
                detail += f" type={el['type']}"
            if el.get('value'):
                detail += f" valeur actuelle=« {el['value']} »"
            lines.append(f"  [{el.get('index')}] {label}{detail}")
    else:
        lines.append('\nAucun élément interactif détecté sur cette page.')

    text = (page.get('text') or '').strip()
    if text:
        lines.append(f'\nTexte visible (extrait) :\n{text}')
    return '\n'.join(lines)


def _render_history(history):
    if not history:
        return 'Aucune action effectuée pour l’instant.'
    lines = []
    for i, entry in enumerate(history, 1):
        outcome = entry.get('outcome') or 'inconnu'
        lines.append(f"  {i}. {entry.get('summary')} → {outcome}")
    return '\n'.join(lines)


def validate_action(action, element_count):
    """Refuse tout ce qui n'est pas exécutable tel quel. Lève PilotError."""
    if not isinstance(action, dict):
        raise PilotError("Réponse de l'IA illisible.")

    tool = action.get('tool')
    if tool not in TOOLS:
        raise PilotError(f"Outil « {tool } » inconnu.")

    args = action.get('args')
    if not isinstance(args, dict):
        args = {}

    missing = TOOLS[tool] - set(args)
    if missing:
        raise PilotError(
            f"Outil « {tool} » : argument(s) manquant(s) : {', '.join(sorted(missing))}."
        )

    if 'index' in TOOLS[tool]:
        index = args.get('index')
        if not isinstance(index, int) or not (0 <= index < element_count):
            raise PilotError(
                f"L'IA a désigné l'élément {index}, hors de la liste "
                f"(0 à {element_count - 1})."
            )

    return {
        'tool': tool,
        # Seuls les arguments attendus sont transmis : rien d'autre n'atteint engine/.
        'args': {k: args[k] for k in TOOLS[tool]},
        'reasoning': (action.get('reasoning') or '').strip(),
    }


def next_action(owner_email, provider, objective, expected_result, page, history):
    """Demande au fournisseur IA la prochaine action, validée et prête à exécuter."""
    message = (
        f'Objectif : {objective}\n'
        f"Résultat attendu : {expected_result or '(non précisé)'}\n\n"
        f'Actions déjà effectuées :\n{_render_history(history)}\n\n'
        f'État de la page :\n{_render_page(page)}'
    )

    raw = ai_client.complete_json(owner_email, provider, SYSTEM, message, ACTION_SCHEMA)
    return validate_action(raw, len(page.get('elements') or []))
